### PR TITLE
Behavior server error codes

### PR DESCRIPTION
**Pull request by jwallace42**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | NONE |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | turtlebot simulation |

---

## Description of contribution in a few bullet points
Updated the behavior server to handle error codes. 

## Description of documentation updates required from your changes
Added new entries in error code table. 

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_behavior_tree/plugins/action/back_up_action.cpp [hunk 1]
```diff
@@ -46,6 +46,24 @@ void BackUpAction::on_tick()
...
+
+BT::NodeStatus BackUpAction::on_aborted()
+{
+  setOutput("error_code_id", result_.result->error_code);
+  return BT::NodeStatus::FAILURE;
+}
+
+BT::NodeStatus BackUpAction::on_cancelled()
+{
+  setOutput("error_code_id", result_.result->error_code);
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/action/back_up_action.cpp:63]: Shouldn't this be none if canceled?

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/spin_action.hpp [hunk 2]
```diff
@@ -55,11 +59,28 @@ class SpinAction : public BtActionNode<nav2_msgs::action::Spin>
       {
         BT::InputPort<double>("spin_dist", 1.57, "Spin distance"),
         BT::InputPort<double>("time_allowance", 10.0, "Allowed time for spinning"),
-        BT::InputPort<bool>("is_recovery", true, "True if recovery")
+        BT::InputPort<bool>("is_recovery", true, "True if recovery"),
+        BT::OutputPort<ActionResult::_error_code_type>(
+            "error_code_id", "The spin behavior error code")
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/spin_action.hpp:64]: New port to the BT.xml file for use + the behavior tree node xml configuration for Groot

File: nav2_msgs/action/Wait.action [hunk 3]
```diff
@@ -1,8 +1,13 @@
+# Error codes
+# Note: The expected priority order of the error should match the message order
+uint16 NONE=0
```
Comment (SteveMacenski) [inline nav2_msgs/action/Wait.action:3]: This feels unnecessary if there's actually no real error to be had here

File: nav2_msgs/action/Spin.action [hunk 4]
```diff
@@ -1,9 +1,18 @@
+# Error codes
+# Note: The expected priority order of the error should match the message order
+uint16 NONE=0
+uint16 UNKNOWN=700
+uint16 TIMEOUT=701
+uint16 TF_ERROR=702
+uint16 COLLISION_AHEAD=703
```
Comment (SteveMacenski) [inline nav2_msgs/action/Spin.action:7]: No invalid input?

File: nav2_behaviors/plugins/drive_on_heading.hpp [hunk 5]
```diff
@@ -83,25 +83,25 @@ class DriveOnHeading : public TimedBehavior<ActionT>
...
+  ResultStatus onCycleUpdate()
   {
     rclcpp::Duration time_remaining = end_time_ - this->steady_clock_.now();
     if (time_remaining.seconds() < 0.0 && command_time_allowance_.seconds() > 0.0) {
       this->stopRobot();
       RCLCPP_WARN(
         this->logger_,
         "Exceeded time allowance before reaching the DriveOnHeading goal - Exiting DriveOnHeading");
-      return Status::FAILED;
+      return ResultStatus{Status::FAILED, 0};
```
Comment (SteveMacenski) [inline nav2_behaviors/plugins/drive_on_heading.hpp:104]: `0` should be a named error for readability

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Generally LGTM, usual docs + updates for xml and groot vis.

File: nav2_msgs/action/Spin.action (hunk 4 above)
Comment (jwallace42) [reply nav2_msgs/action/Spin.action:7]: Spin currently has no check to invalidate any input.

File: nav2_msgs/action/Wait.action [hunk 6]
```diff
@@ -3,6 +3,7 @@ builtin_interfaces/Duration time
 ---
 #result definition
 builtin_interfaces/Duration total_elapsed_time
+uint16 error_code
```
Comment (SteveMacenski) [inline nav2_msgs/action/Wait.action:6]: This feels unnecessary if there's no error codes

File: nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp [hunk 7]
```diff
@@ -45,6 +45,12 @@ enum class Status : int8_t
   RUNNING = 3,
 };
 
+struct ResultStatus
+{
+  Status status;
+  unsigned short error_code{0};  //NOLINT
```
Comment (SteveMacenski) [inline nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp:51]: Why does this need no linting?

**Review by SteveMacenski — COMMENTED** [review]
I think last is for the BTs to use it. Anything else you think we need? Maybe I suppose unit tests for the new functions added if not already covered

File: nav2_msgs/action/Wait.action (hunk 6 above)
Comment (jwallace42) [reply nav2_msgs/action/Wait.action:6]: Yeah, but anything using the timed behavior needs a error code.

File: nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp (hunk 7 above)
Comment (SteveMacenski) [reply nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp]: ^ not much of a reason for this not to just be in unsigned int, really. But I don't really care

**Comment by SteveMacenski** [comment]
Just on CI

File: nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp (hunk 7 above)
Comment (jwallace42) [reply nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp]: Switched this to a uint16_t :)

**Comment by jwallace42** [comment]
Sweet. 
 
I will finish those up :)

**Comment by SteveMacenski** [comment]
Should the default BTs be updated to use them? I suppose not necessarily required, but might be good at least for the main 2-3 trees since those are minimal examples folks will work from

**Comment by SteveMacenski** [comment]
This is actually a time when mergify is correct, rebase :smile:

**Review by SteveMacenski — APPROVED** [review]
Waiting on CI. I'll look tomorrow

**Pull request merged**